### DISCUSSION TITLE
Emqx production deployment

### DISCUSSION THREAD

**Discussion by alexa22013** [emqx/emqx#11507] [General]
When deploying emqx v5 on production is there and recommendation as to expose only replicants or both core and replicants for client connection. This is to understand why the emqx listener service created by emqx operator are only for replicant nodes. 

Also is there any recommendations based on having more publishers than subcribers and vice versa.

Thanks in advance!

**Comment by kjellwinblad** [comment]
I think we recommend that clients only connect to replicant nodes in a cluster setup so that the core nodes have more resources to do message routing (@ieQu1 you can correct me if I'm wrong).
 
Regarding "having more publishers than subscribers and vice versa", I think that have to depend on the type of system you are building. I don't think we have any specific recommendation regarding the publisher/subscriber ratio.

**Reply by alexa22013** [reply]
@ieQu1 any take on this?

**Reply by ieQu1** [reply]
Core and replicant nodes run the same code, and are capable of routing messages, running rules, dashboard, etc. They are more or less equivalent functionally.

The difference between the roles is rather subtle: internally, EMQX uses a database called [Mria](https://github.com/emqx/mria), which stores various persistent and runtime information, including the routing table. The node role (core/replicant) determines how Mria operates on the node:

- Core nodes can write data to the internal tables. When the client subscribes to a topic filter, for example, core node creates a few entries in the internal Mria tables related to the topics and the routes.  
- Replicant nodes have read-only replicas of the tables. For example, they can read the routing table, but they cannot update it directly; so they delegate this operation to one of the core nodes instead. 

As you can see, core nodes have to do extra work: they have to accept transaction requests from the replicants, and also push the updates back. Therefore, it's important that they remain responsive regardless of the volume of MQTT traffic; otherwise performance of the whole cluster will get a hit.

So, while the core nodes _can_ serve the clients, in the large clusters with a lots of replicants it's best to avoid that.

Second question:

> Also is there any recommendations based on having more publishers than subcribers and vice versa.

There are no general rules, and everything depends on the use case. We optimize performance for both cases.
